File: app/Modules/Gateway/Http/Controllers/WebhookController.php
<?php

namespace App\Modules\Gateway\Http\Controllers;

use App\Modules\Gateway\Models\Transaction;
use App\Modules\Gateway\Payments\PaymentProviderInterface;
use App\Modules\Gateway\Payments\WebhookResult;
use App\Modules\Gateway\Services\TransactionService;
use Illuminate\Http\Request;

class WebhookController extends Controller
{
    public function __construct(
        private readonly PaymentProviderInterface $provider,
        private readonly TransactionService $transactions,
    ) {
    }

    /**
     * POST /webhooks/payu — notyfikacje PayU. Po poprawnej weryfikacji
     * podpisu zawsze odpowiadamy 200 (inaczej PayU ponawia).
     */
    public function payu(Request $request)
    {
        $result = $this->provider->handleWebhook($request);

        if (! $result->valid) {
            return response()->json(['error' => 'Niepoprawny podpis'], 401);
        }

        if ($result->transactionId) {
            $transaction = Transaction::find($result->transactionId);

            if ($transaction) {
                match ($result->status) {
                    WebhookResult::STATUS_PAID => $this->transactions->markPaid($transaction),
                    WebhookResult::STATUS_FAILED => $this->transactions->markFailed($transaction),
                    default => null,
                };
            }
        }

        return response()->json(['ok' => true]);
    }
}

File: pay-backend/gateway-api/src/main/kotlin/com/supportme/pay/gateway/api/payment/WebhookController.kt
package com.supportme.pay.gateway.api.payment

import com.supportme.pay.gateway.domain.repository.TransactionRepository
import com.supportme.pay.gateway.payments.PaymentProvider
import com.supportme.pay.gateway.payments.WebhookResult
import jakarta.servlet.http.HttpServletRequest
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.UUID

/**
 * Odpowiednik `WebhookController::payu` — notyfikacje PayU. Po poprawnej
 * weryfikacji podpisu ZAWSZE 200 (inaczej PayU ponawia w nieskończoność),
 * nawet gdy transakcja nie zostanie znaleziona.
 */
@RestController
@RequestMapping("/webhooks")
class WebhookController(
    private val paymentProvider: PaymentProvider,
    private val transactionRepository: TransactionRepository,
    private val transactionService: TransactionService,
) {

    @PostMapping("/payu")
    fun payu(request: HttpServletRequest): ResponseEntity<Map<String, Any>> {
        val rawBody = request.reader.readText()
        val signatureHeader = request.getHeader("OpenPayu-Signature")

        val result = paymentProvider.handleWebhook(rawBody, signatureHeader)
        if (!result.valid) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(mapOf("error" to "Niepoprawny podpis"))
        }

        result.transactionId?.let { txId ->
            runCatching { UUID.fromString(txId) }.getOrNull()?.let { uuid ->
                transactionRepository.findById(uuid).ifPresent { transaction ->
                    when (result.status) {
                        WebhookResult.Status.PAID -> transactionService.markPaid(transaction)
                        WebhookResult.Status.FAILED -> transactionService.markFailed(transaction)
                        WebhookResult.Status.IGNORED -> Unit
                    }
                }
            }
        }

        return ResponseEntity.ok(mapOf("ok" to true))
    }
}
